draft-yasskin-http-origin-signed-responses.md

```diff
@@ -0,0 +1,316 @@
+
+author:
+ -
+    name: Jeffrey Yasskin
+    organization: Google
+    email: jyasskin@chromium.org
+
+informative:
+  SRI: W3C.REC-SRI-20160623
+
+--- abstract
+
+This document explores how a server can send particular responses that are
+authoritative for an origin, when the server itself is not authoritative for
+that origin. For now, it focuses on the constraints covering any such mechanism.
+
+--- note_Note_to_Readers
+
+Discussion of this draft takes place on the HTTP working group mailing list
+(ietf-http-wg@w3.org), which is archived
+at <https://lists.w3.org/Archives/Public/ietf-http-wg/>.
+
+The source code and issues list for this draft can be found
+in <https://github.com/WICG/webpackage>.
+
+--- middle
+
+# Introduction
+
+When
+I
+[presented Web Packaging to DISPATCH](https://datatracker.ietf.org/doc/minutes-99-dispatch/),
+folks thought it would make sense to split it into a way to sign individual HTTP
+responses as coming from a particular origin, and separately a way to bundle a
+collection of HTTP responses. This document explores the constraints on any
+method of signing HTTP responses and briefly sketches a possible solution to the
+constraints.
+
+# Terminology
+
+Author
+: The entity that controls the server for a particular origin {{?RFC6454}}. The
+  author can get a CA to issue certificates for their private keys and can run a
+  TLS server for their origin.
+
+Intermediate
+: An entity that fetches signed HTTP responses from an author or another
+  intermediate and forwards them to another intermediate or a client.
+
+Client
+: An entity that uses a signed HTTP response and needs to be able to prove that
+  the author vouched for it as coming from its claimed origin.
+
+# Use cases
+
+## PUSHed subresources {#uc-pushed-subresources}
+
+To reduce round trips, a server might use HTTP/2 PUSH to inject a subresource
+from another server into the client's cache. If anything about the subresource
+is expired or can't be verified, the client would fetch it from the original
+server.
+
+For example, if `https://example.com/index.html` includes
+
+~~~html
+<script src="https://jquery.com/jquery-1.2.3.min.js">
+~~~
+
+Then to avoid the need to look up and connect to `jquery.com` in the critical
+path, `example.com` might PUSH that resource ({{?RFC7540}}, section 8.2), signed
+by `jquery.com`.
+
+## Explicit use of a CDN for subresources {#uc-explicit-cdn}
+
+In order to speed up loading but still maintain control over its content, an
+HTML page in a particular origin `O.com` could tell clients to load its
+subresources from an intermediate CDN that's not authoritative, but require that
+those resources be signed by `O.com` so that the CDN couldn't modify the
+resources.
+
+~~~html
+<img logicalsrc="https://O.com/img.png"
+     physicalsrc="https://cdn.com/O.com/img.png">
+~~~
+
+This could be used for some of the same purposes as SRI ({{uc-sri}}).
+
+## Subresource Integrity {#uc-sri}
```

Comment:
This may be out of scope, but.. is content-addressable caching something to consider? You note that we want to sign headers+body, but what about cases where I only care about body -- e.g. I don't particularly care which CDN you fetch jquery.v.1.2.3.js from, as long as its content signature matches.

Reply:
There are two separable features here:

1. Content-addressed caching where the resource original author is willing to vouch for the resource's actual URL. That is, if you have content with the same logical URL and a good signature, you can use it without contacting the CDN. (But this isn't actually content-addressing; it's URL-addressing again.) I think this is safe from a same-origin perspective, but I'm not sufficiently current with the SRI folks' discussion of content-addressing to know if it solves their privacy worries. @mikewest?
2. Cases where you only care about the content, and not its URL. I think this doesn't make sense for signed content: the point of a public keypair is that you can sign more than one thing with it, but then which data signed with the provided public key are you actually trying to use?

So ... I'm happy to consider content-addressing, but I think I don't understand what you mean by it in this context.

Reply:
My question is mostly as an exercise to identify whether this is in or out of scope of this proposal. Specifically, you call out signing headers+body, but signing body only may have merit in context of sharing responses -- that said, you're right in this case you cant be signing with your key, so perhaps that's out of scope.

Reply:
I've added a link to https://github.com/mikewest/signature-based-sri/issues/5 arguing that it's unsafe to sign just-a-body, with any key. :smiling_imp: